File: doc-examples/jdbc-vector-example-kotlin/src/main/kotlin/example/DocumentEmbeddingRepository.kt
package example

import io.micronaut.data.annotation.Repository
import io.micronaut.data.jdbc.annotation.JdbcRepository
import io.micronaut.data.model.query.builder.sql.Dialect
import io.micronaut.data.model.vector.Vector
import io.micronaut.data.model.vector.search.Score
import io.micronaut.data.model.vector.search.ScoringFunction
import io.micronaut.data.model.vector.search.SearchResults
import io.micronaut.data.repository.CrudRepository

@JdbcRepository(dialect = Dialect.ORACLE)
@Repository
interface DocumentEmbeddingRepository : CrudRepository<DocumentEmbedding, Long> {

    // Example showing vector as a parameter
    fun save(embedding: Vector): DocumentEmbedding

    fun findTop2ByEmbeddingNear(vec: Vector, maxDistance: Double): List<DocumentEmbedding>

    fun searchByEmbeddingNear(vector: Vector,
                              maxDistance: Score,
                              function: ScoringFunction): SearchResults<DocumentEmbedding>
}

File: doc-examples/jdbc-vector-example-python/src/main/python/example/DocumentEmbeddingRepository.py
from micronaut.data.annotation import Repository
from micronaut.data.jdbc.annotation import JdbcRepository
from micronaut.data.model.query.builder.sql import Dialect
from micronaut.data.model.vector import Vector
from micronaut.data.model.vector.search import Score, ScoringFunction, SearchResults
from micronaut.data.repository import CrudRepository

from example.DocumentEmbedding import DocumentEmbedding


@JdbcRepository(dialect=Dialect.ORACLE)
@Repository
class DocumentEmbeddingRepository(CrudRepository[DocumentEmbedding, int]):

    # Example showing vector as a parameter
    def save(self, embedding: Vector) -> DocumentEmbedding: ...

    def findTop2ByEmbeddingNear(self, vec: Vector, maxDistance: float) -> list[DocumentEmbedding]: ...

    def searchByEmbeddingNear(self, vector: Vector, maxDistance: Score, function: ScoringFunction) -> SearchResults[DocumentEmbedding]: ...
